esphome/components/display/display_buffer.cpp

```diff
@@ -773,12 +773,36 @@ Color Animation::get_grayscale_pixel(int x, int y) const {
   return Color(gray, gray, gray, alpha);
 }
 Animation::Animation(const uint8_t *data_start, int width, int height, uint32_t animation_frame_count, ImageType type)
-    : Image(data_start, width, height, type), current_frame_(0), animation_frame_count_(animation_frame_count) {}
+    : Image(data_start, width, height, type),
+      current_frame_(0),
+      animation_frame_count_(animation_frame_count),
+      loop_start_frame_(0),
+      loop_end_frame_(animation_frame_count_),
+      loop_count_(0),
+      loop_current_iteration_(1) {}
 int Animation::get_animation_frame_count() const { return this->animation_frame_count_; }
 int Animation::get_current_frame() const { return this->current_frame_; }
+void Animation::set_loop(int start_frame, int end_frame, int count) {
+  loop_start_frame_ = std::min(start_frame, animation_frame_count_);
+  if (end_frame == -1) {
+    loop_end_frame_ = animation_frame_count_;
+  } else {
+    loop_end_frame_ = std::min(end_frame, animation_frame_count_);
+  }
+  loop_count_ = count;
+}
 void Animation::next_frame() {
   this->current_frame_++;
+  if (loop_count_ && this->current_frame_ <= loop_end_frame_) {
+    if (this->loop_current_iteration_ < loop_count_ || loop_count_ < 0) {
+      if (this->current_frame_ >= loop_end_frame_) {
+        this->current_frame_ = loop_start_frame_;
+        this->loop_current_iteration_++;
+      }
+    }
+  }
```

Comment:
Your innermost `if` is effectively a `==` as you already checked for `<=` in the outermost `if`.
You can merge that into one or two `if`s:

```suggestion
  if (loop_count_ && this->current_frame_ == loop_end_frame_
      && (this->loop_current_iteration_ < loop_count_ || loop_count_ < 0)) {
    this->current_frame_ = loop_start_frame_;
    this->loop_current_iteration_++;
  }
```

Reply:
Thanks for pointing this out. I spent so much time trying to make the loop work at all, that I did not see the redundancy anymore.

Reply:
Happens to all of us 😄 That's exactly what code reviews are for.